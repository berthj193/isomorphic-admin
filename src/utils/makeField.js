import React from 'react';
import PropTypes from 'prop-types';
import { FormItem } from '../components/uielements/form';

const makeField = (Component, options = {}) =>
  class CustomComponent extends React.Component {
    static propTypes = {
      label: PropTypes.string,
      meta: PropTypes.shape({
        touched: PropTypes.bool,
        invalid: PropTypes.invalid,
      }),
      hasFeedback: PropTypes.bool,
      children: PropTypes.node,
      input: PropTypes.object,
      required: PropTypes.bool,
    };

    formItemLayoutWithoutLabel = {
      wrapperCol: {
        xs: { offset: 24, span: 24 },
        sm: { offset: 4, span: 6 },
      },
    };

    render() {
      const { withoutLabel } = options;
      const { meta = {}, label, hasFeedback, input, children, required, ...rest } = this.props;
      const hasError = meta.touched && meta.invalid;

      if (withoutLabel) {
        return <FormItem
          {...this.formItemLayoutWithoutLabel}
          validateStatus={hasError ? 'error' : 'success'}
          hasFeedback={hasFeedback && hasError}
          help={hasError && meta.error}
          required={required}
        >
          <Component {...input} {...rest}>{children}</Component>
        </FormItem>;
      } else {
        return (
          <FormItem
            label={label}
            validateStatus={hasError ? 'error' : 'success'}
            hasFeedback={hasFeedback && hasError}
            help={hasError && meta.error}
            required={required}
          >
            <Component {...input} {...rest}>{children}</Component>
          </FormItem>
        );
      }
    }
  };

makeField.displayName = 'makeField';

export default makeField;
