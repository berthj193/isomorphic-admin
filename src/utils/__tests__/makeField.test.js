import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import makeField from '../makeField';

describe('makeField util', () => {
  const textField = (props = {}) => <input {...props} />;

  it('should render component without label properly', () => {
    const Component = makeField(textField, {
      withoutLabel: true,
    });
    expect(toJson(shallow(<Component />))).toMatchSnapshot();
  });

  it('should render component without label with correct props', () => {
    const Component = makeField(textField, {
      withoutLabel: true,
    });
    expect(toJson(shallow(
      <Component
        input={{ inputProp: 'foo' }}
        anotherProp="anotherProp"
      />
    ))).toMatchSnapshot();
  });

  it('should render component with label properly', () => {
    const Component = makeField(textField);
    expect(toJson(shallow(<Component label="foo" />))).toMatchSnapshot();
  });

  it('should render component with required prop', () => {
    const Component = makeField(textField);
    expect(toJson(shallow(<Component label="foo" required />))).toMatchSnapshot();
  });

  it('should render component with an error', () => {
    const Component = makeField(textField);
    expect(toJson(shallow(
      <Component
        label="foo"
        meta={{
          error: 'required',
          touched: true,
          invalid: true,
        }}
        required
      />
    ))).toMatchSnapshot();
  });

  it('should render component with error feedback icon', () => {
    const Component = makeField(textField);
    expect(toJson(shallow(
      <Component
        label="foo"
        meta={{
          error: 'required',
          touched: true,
          invalid: true,
        }}
        hasFeedback
        required
      />
    ))).toMatchSnapshot();
  });
});
