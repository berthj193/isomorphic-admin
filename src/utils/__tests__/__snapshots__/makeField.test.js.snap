// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeField util should render component with an error 1`] = `
<FormItem
  help="required"
  label="foo"
  required={true}
  validateStatus="error"
>
  <textField />
</FormItem>
`;

exports[`makeField util should render component with error feedback icon 1`] = `
<FormItem
  hasFeedback={true}
  help="required"
  label="foo"
  required={true}
  validateStatus="error"
>
  <textField />
</FormItem>
`;

exports[`makeField util should render component with label properly 1`] = `
<FormItem
  label="foo"
  validateStatus="success"
>
  <textField />
</FormItem>
`;

exports[`makeField util should render component with required prop 1`] = `
<FormItem
  label="foo"
  required={true}
  validateStatus="success"
>
  <textField />
</FormItem>
`;

exports[`makeField util should render component without label properly 1`] = `
<FormItem
  validateStatus="success"
  wrapperCol={
    Object {
      "sm": Object {
        "offset": 4,
        "span": 6,
      },
      "xs": Object {
        "offset": 24,
        "span": 24,
      },
    }
  }
>
  <textField />
</FormItem>
`;

exports[`makeField util should render component without label with correct props 1`] = `
<FormItem
  validateStatus="success"
  wrapperCol={
    Object {
      "sm": Object {
        "offset": 4,
        "span": 6,
      },
      "xs": Object {
        "offset": 24,
        "span": 24,
      },
    }
  }
>
  <textField
    anotherProp="anotherProp"
    inputProp="foo"
  />
</FormItem>
`;
